server/src/server.ts

```diff
@@ -394,6 +398,55 @@ function parseFile(textDocUri: string, docContent: string) {
         connection.sendDiagnostics({ uri: textDocUri, diagnostics: [] });
     }
 }
+connection.onDocumentSymbol((_params: DocumentSymbolParams) => {
+    return getDocumentSymbols(_params.textDocument.uri);
+});
+
+/**
+ * Build the list of all available YML symbols that are in a file.
+ *
+ * @param uri the file URI
+ *
+ * @returns the list of available YML symbols.
+ */
+function getDocumentSymbols(uri: string) {
+    // Get the YML objects definitions that are in this file.
+    let currentDocSymbols = definitionsProvider.definitions.filter(
+        (elem) => elem.uri === uri && !!elem.definitionLocation,
+    );
+    // Also get the YML objects implementations that are in this file.
+    currentDocSymbols = currentDocSymbols.concat(
+        definitionsProvider.implementations.filter((elem) => elem.uri === uri && !!elem.definitionLocation),
+    );
+
+    const parentToChildren = new Map<AbstractYmlObject, AbstractYmlObject[]>();
+    for (const elem of currentDocSymbols) {
+        if (!!elem.sourceElement && !parentToChildren.has(elem.sourceElement)) {
+            // Current element has a parent, but we don't have an entry for it yet.
+            connection.console.info(`setting entry for key ${elem.sourceElement} because it has a child.`);
+            parentToChildren.set(elem.sourceElement, [elem]);
+        } else if (!!elem.sourceElement) {
+            // Current element has a parent and we already have an entry for it.
+            parentToChildren.get(elem.sourceElement).push(elem);
+        } else if (!parentToChildren.has(elem)) {
+            // Current element has no parent and this is the first time we encounter it.
+            connection.console.info(`resetting entry for key ${elem.label}.`);
+            parentToChildren.set(elem, []);
+        } else {
+            // Current element already exists in the map and has no parent.
+            // Basically, this means that this element is already set as the parent of another element.
+        }
+    }
+
+    const candidates: DocumentSymbol[] = [];
```

Comment:
> See what I mean?

Yes. I'll rename this variable, then.

> By the way, don’t some guys appear several times in your structure?

It could, but it doesn't. The reason is that there are only three parents: classes, functions and enums.
There are no elements that are parent and child at the same time.